src/Suave/Web.fs

```diff
@@ -278,20 +277,37 @@ module ParsingAndControl =
 
   /// Process the request, reading as it goes from the incoming 'stream', yielding a HttpRequest
   /// when done
-  let process_request is_secure (ctx : HttpContext) : SocketOp<HttpContext option> = socket {
+  let process_request ({ connection = connection
+                         runtime    = runtime } as ctx)
+                      : SocketOp<HttpContext option> = socket {
 
-    let! (first_line : string), connection = read_line ctx.connection
+    let! (first_line : string), connection' = read_line connection
 
-    let raw_method, url, raw_query, http_version = parse_url first_line
+    let raw_method, path, raw_query, http_version = parse_url first_line
     let meth = HttpMethod.parse raw_method
 
-    let! headers, connection = read_headers connection
+    let! headers, connection'' = read_headers connection'
 
-    let request =
-      HttpRequest.mk http_version url meth headers raw_query
-        (parse_trace_headers headers) is_secure connection.ipaddr
+    // TODO: if client Host header not present, respond with 400 Bad Request as
+    // per http://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html
+    let host =
+      headers %% "host" |> Option.get
+      |> function
+      | s when System.Text.RegularExpressions.Regex.IsMatch(s, ":\d+$") ->
```

Comment:
Would be nice with real parsers like Freya has, for all of this stuff, so that not both the regex and then the substring has to match.